Kotlin daemon JVM args - clearing up inheritance and defaults (#4454)
JetBrains/kotlin-web-site: docs/topics/gradle/gradle-compilation-and-caches.md

#### Gradle daemon arguments inheritance

> The Kotlin daemon inherits a specific set of arguments from the Gradle daemon and then overwrites them with JVM args specified for the Kotlin daemon. For example, in the `gradle.properties` file:

Reviewer: ```suggestion
By default, the Kotlin daemon inherits a specific set of arguments from the Gradle daemon but overwrites them with any 
JVM arguments specified directly for the Kotlin daemon. For example, if you add the following JVM arguments in the `gradle.properties` file:
```

```none
org.gradle.jvmargs=-Xmx1500m -Xms500m -XX:MaxMetaspaceSize=1g
```

> would then result in the Kotlin daemon's JVM args being:

Reviewer 2: ```suggestion
would then result in the Kotlin daemon's JVM args being similar to:
```

> would then result in the Kotlin daemon's JVM args being similar to:

Reviewer: ```suggestion
These arguments are then added to the Kotlin daemon's JVM arguments:
```

```none
-Xmx1500m -XX:ReservedCodeCacheSize=320m -XX:MaxMetaspaceSize=1g -XX:UseParallelGC -ea -XX:+UseCodeCacheFlushing -XX:+HeapDumpOnOutOfMemoryError -Djava.awt.headless=true -Djava.rmi.server.hostname=127.0.0.1 --add-exports=java.base/sun.nio.ch=ALL-UNNAMED
```

> > Learn more about [Kotlin daemon's behavior with JVM arguments](#kotlin-daemon-s-behavior-with-jvm-arguments).

Reviewer: ```suggestion
> To learn more about the Kotlin daemon's default behavior with JVM arguments, see [Kotlin daemon's behavior with JVM arguments](#kotlin-daemon-s-behavior-with-jvm-arguments).
```

#### kotlin.daemon.jvmargs property

> You can add the `kotlin.daemon.jvmargs` property in the `gradle.properties` file. Note that if you do not specify `ReservedCodeCacheSize` here or in Gradle's JVM args, a default will be applied by the Kotlin Gradle plugin:

Reviewer: ```suggestion
You can add the `kotlin.daemon.jvmargs` property in the `gradle.properties` file:
```

```none
kotlin.daemon.jvmargs=-Xmx1500m -Xms500m
```

> ```none

Reviewer: ```suggestion
Note that if you don't specify the `ReservedCodeCacheSize` argument here or in Gradle's JVM arguments, the Kotlin Gradle plugin applies a default value of `320m`:

```none
```
I think that it's better to have this sentence and it's sample separate from the first. Otherwise the second sample looks lost.

  > even if other requested JVM arguments are different, this daemon will be reused instead of starting a new one.
  >
  {style="note"}

> If the following arguments are not specified, the Kotlin daemon will inherit them from the Gradle daemon:

Reviewer: ```suggestion
If the following arguments aren't specified, the Kotlin daemon inherits them from the Gradle daemon:
```

> * -Xmx
> * -XX:MaxMetaspaceSize
> * -XX:ReservedCodeCacheSize (Defaults to 320m if not inherited and unspecified)

Reviewer: ```suggestion
* `-Xmx`
* `-XX:MaxMetaspaceSize`
* `-XX:ReservedCodeCacheSize`. If not specified or inherited, the default value is `320m`.
```

> The Kotlin daemon also has a number of defaults:

Reviewer: ```suggestion
The Kotlin daemon has the following default JVM arguments:
```

> * -XX:UseParallelGC (if no other garbage collector was set)
> * -ea
> * -XX:+UseCodeCacheFlushing
> * -Djava.awt.headless=true
> * -D{java.servername.property}={localhostip}
> * --add-exports=java.base/sun.nio.ch=ALL-UNNAMED (Only for JDK 16+)

Reviewer: ```suggestion
* `-XX:UseParallelGC`. This argument is only applied if no other garbage collector is specified.
* `-ea`
* `-XX:+UseCodeCacheFlushing`
* `-Djava.awt.headless=true`
* `-D{java.servername.property}={localhostip}`
* `--add-exports=java.base/sun.nio.ch=ALL-UNNAMED`. This argument is only applied for JDK versions 16 or higher.
```

> The list of defaults is considered an implementation detail of the Kotlin daemon and may vary between versions. You can use a tool like [VisualVM](https://visualvm.github.io/) to verify what the actual settings of a running JVM process like the Kotlin daemon are.

Reviewer: ```suggestion
> The list of default JVM arguments for the Kotlin daemon may vary between versions. You can use a tool like [VisualVM](https://visualvm.github.io/) to check the actual settings of a running JVM process, like the Kotlin daemon.
>
{style="note"}
```

  > even if other requested JVM arguments are different, this daemon will be reused instead of starting a new one.
  >
  {type="note"}

If the following arguments are not specified, the Kotlin daemon will inherit them from the Gradle daemon:

* -Xmx
* -XX:MaxMetaspaceSize
* -XX:ReservedCodeCacheSize (Defaults to 320m if not inherited and unspecified)

The Kotlin daemon also has a number of defaults:
* -XX:UseParallelGC (if no other garbage collector was set)
* -ea
* -XX:+UseCodeCacheFlushing

> * -XX:+HeapDumpOnOutOfMemoryError

Reviewer 2: This one have to be configured by your JVM or some other implicit configuration like the `JAVA_TOOL_OPTIONS` environment variable. That's not what does KGP configure by default.

Author: Ah, my mistake, will remove

* -Djava.awt.headless=true
* -D{java.servername.property}={localhostip}
* --add-exports=java.base/sun.nio.ch=ALL-UNNAMED (Only for JDK 16+)

> You can use a tool like [VisualVM](https://visualvm.github.io/) to verify what the actual settings of a running JVM process like the Kotlin daemon are.

Reviewer 2: Could you rephrase this as 
```suggestion
The list of defaults is considered an implementation detail of the Kotlin daemon and may vary between versions. You can use a tool like [VisualVM](https://visualvm.github.io/) to verify what the actual settings of a running JVM process like the Kotlin daemon are.
```
